### PR TITLE
e2e: add etcd backup operator test

### PR DESCRIPTION
**Pull request by fanminshi**
Build unified image in build_pass allowing e2e test to run for backup_operator from the image.

Add an e2e test for etcd backup operator:

1. spawn etcd backup operator pod.
2. create etcd cluster.
3. invoke a backup cr for the etcd cluster.
4. verify that backup status shows succeeded.

### CODE REVIEW

**Comment by fanminshi** [comment]
via https://jenkins-etcd-public.prod.coreos.systems/job/etcd-operator-e2eslow-pr/567/console

```
19:29:19   Normal   Pulling                19s (x3 over 1m)  kubelet, gke-jenkins-default-pool-0a70060f-rfh0  pulling image "gcr.io/coreos-k8s-scale-testing/etcd-operator:69627235f569f9359ef72c7161f4e82a3b65c68b"
19:29:19   Warning  Failed                 19s (x3 over 1m)  kubelet, gke-jenkins-default-pool-0a70060f-rfh0  Failed to pull image "gcr.io/coreos-k8s-scale-testing/etcd-operator:69627235f569f9359ef72c7161f4e82a3b65c68b": rpc error: code = 2 desc = Error: Status 405 trying to pull repository coreos-k8s-scale-testing/etcd-operator: "v1 Registry API is disabled. If you are not explicitly using the v1 Registry API, it is possible your v2 image could not be found. Verify that your image is available, or retry with `dockerd --disable-legacy-registry`. See https://cloud.google.com/container-registry/docs/support/deprecation-notices"
19:29:19   Warning  FailedSync             6s (x6 over 1m)   kubelet, gke-jenkins-default-pool-0a70060f-rfh0  Error syncing pod
19:29:19   Normal   BackOff                6s (x3 over 59s)  kubelet, gke-jenkins-default-pool-0a70060f-rfh0  Back-off pulling image "gcr.io/coreos-k8s-scale-testing/etcd-operator:69627235f569f9359ef72c7161f4e82a3b65c68b"
19:29:19 " 
19:29:19 time="2017-10-11T19:29:19Z" level=error msg="fail to setup framework: failed to setup etcd operator: failed to wait pod running, it is still pending: still failing after 12 retries" 
```

failed to pull image again.

**Comment by fanminshi** [comment]
@etcd-bot test this please
reason: jenkins internal error
```
00:10:36 FATAL: command execution failed
00:10:36 java.io.IOException
00:10:36 	at hudson.remoting.Channel.close(Channel.java:1292)
```

**Comment by fanminshi** [comment]
cc/ @hongchaodeng @hasbro17

**Comment by hongchaodeng** [comment]
I think we should create a PR first to make backup operator create CRD.

**Comment by fanminshi** [comment]
@hongchaodeng will separate this pr into many sub prs.

**Comment by fanminshi** [comment]
PTAL cc/ @hongchaodeng

File: test/e2e/framework/framework.go [hunk 1]
```diff
@@ -101,7 +100,6 @@ func (f *Framework) setup() error {
 	if err := f.SetupEtcdOperator(); err != nil {
 		return fmt.Errorf("failed to setup etcd operator: %v", err)
 	}
-	logrus.Info("etcd operator created successfully")
```
Comment (hongchaodeng) [inline test/e2e/framework/framework.go]: Keep this?

File: test/e2e/backup_test.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TestEtcdBackupOperatorForS3Backup tests if etcd backup operator can save etcd backup to S3.
+func TestEtcdBackupOperatorForS3Backup(t *testing.T) {
+	if os.Getenv(envParallelTest) == envParallelTestTrue {
+		t.Parallel()
+	}
+	f := framework.Global
+	if err := f.SetupEtcdBackupOperator(); err != nil {
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: Don't create it here. Keep the lifecycle in main_test.go

File: test/e2e/framework/framework.go (hunk 1 above)
Comment (fanminshi) [reply test/e2e/framework/framework.go]: yep

**Comment by fanminshi** [comment]
@hongchaodeng where do you think i should add more docs?

**Comment by hongchaodeng** [comment]
docs for public methods

**Comment by fanminshi** [comment]
added doc.

File: test/e2e/framework/framework.go [hunk 3]
```diff
@@ -191,6 +199,54 @@ func (f *Framework) DeleteEtcdOperatorCompletely() error {
 	return nil
 }
 
+const etcdBackupOperator = "etcd-backup-operator"
```
Comment (hongchaodeng) [inline test/e2e/framework/framework.go]: Put const at top

File: test/e2e/backup_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		t.Fatalf("failed to create etcd backup cr: %v", err)
+	}
+	defer func() {
+		if err := e2eutil.DeleteBackupCR(t, f.CRClient, f.KubeClient, eb); err != nil {
+			t.Fatalf("failed to delete etcd backup cr: %v", err)
+		}
+	}()
+
+	err = retryutil.Retry(time.Second, 10, func() (bool, error) {
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: Why retry **every second** and **10 times**?

File: test/e2e/backup_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		reb, err := f.CRClient.EtcdV1beta2().EtcdBackups(f.Namespace).Get(eb.Name, v1.GetOptions{})
+		if err != nil {
+			return true, fmt.Errorf("failed to retrieve backup CR: %v", err)
+		}
+		if reb.Status.Succeeded {
+			return true, nil
+		}
+		return false, nil
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: Add a TODO: also check if backup is there on S3 path.

File: test/e2e/backup_test.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	defer func() {
+		if err := e2eutil.DeleteCluster(t, f.CRClient, f.KubeClient, testEtcd); err != nil {
+			t.Fatal(err)
+		}
+	}()
+	if _, err := e2eutil.WaitUntilSizeReached(t, f.CRClient, 3, 6, testEtcd); err != nil {
+		t.Fatalf("failed to create 3 members etcd cluster: %v", err)
+	}
+
+	eb, err := e2eutil.CreateBackupCR(t, f.CRClient, f.Namespace, e2eutil.NewS3Backup(testEtcd.Name))
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: No need to abstract a method here.

File: test/e2e/backup_test.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	if err != nil {
+		t.Fatalf("failed to create etcd backup cr: %v", err)
+	}
+	defer func() {
+		if err := e2eutil.DeleteBackupCR(t, f.CRClient, f.KubeClient, eb); err != nil {
```
Comment (hongchaodeng) [inline test/e2e/backup_test.go]: Same. No need to abstract a method here.

File: test/e2e/backup_test.go (hunk 4 above)
Comment (hongchaodeng) [reply test/e2e/backup_test.go]: What's the usual case for manual retry?

File: test/e2e/backup_test.go (hunk 6 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: which method? `CreateBackupCR` or `NewS3Backup`? 
I think CreateBackupCR can be useful if we are going to add more tests that needs to create backup cr.

Comment (hongchaodeng) [reply test/e2e/backup_test.go]: > which method? CreateBackupCR or NewS3Backup?

CreateBackupCR

> I think CreateBackupCR can be useful if we are going to add more tests that needs to create backup cr.

The client method is also simple enough.

File: test/e2e/backup_test.go (hunk 4 above)
Comment (fanminshi) [reply test/e2e/backup_test.go]: @hongchaodeng 10 seconds is an arbitrary timeout. I don't remember the usually waiting time for manual case. The point of retry is just a 10 seconds timeout. I suspect if in 10 seconds, I don't a snapshot response. something is wrong for sure.

**Comment by fanminshi** [comment]
PTAL cc/ @hongchaodeng

**Comment by hongchaodeng** [comment]
@fanminshi 
Let's pause this and implement status.S3Path first?

File: test/e2e/framework/framework.go [hunk 8]
```diff
@@ -191,6 +201,52 @@ func (f *Framework) DeleteEtcdOperatorCompletely() error {
...
+	if err != nil {
+		return err
+	}
+	logrus.Infof("etcd backup operator pod is running on node (%s)", p.Spec.NodeName)
+
+	return e2eutil.WaitUntilOperatorReady(f.KubeClient, f.Namespace, etcdBackupOperator)
+}
+
+// DeleteEtcdBackupOperator deletes an etcd backup operator with the name as "etcd-backup-operator".
+func (f *Framework) DeleteEtcdBackupOperator() error {
```
Comment (hongchaodeng) [inline test/e2e/framework/framework.go]: No need to define this method

File: test/e2e/framework/framework.go [hunk 9]
```diff
@@ -191,6 +201,52 @@ func (f *Framework) DeleteEtcdOperatorCompletely() error {
...
+		},
+	}
+
+	p, err := k8sutil.CreateAndWaitPod(f.KubeClient, f.Namespace, pod, 60*time.Second)
+	if err != nil {
+		return err
+	}
+	logrus.Infof("etcd backup operator pod is running on node (%s)", p.Spec.NodeName)
+
+	return e2eutil.WaitUntilOperatorReady(f.KubeClient, f.Namespace, etcdBackupOperator)
```
Comment (hongchaodeng) [inline test/e2e/framework/framework.go]: Just return nil.
There is no readiness check in backup operator yet.

Comment (fanminshi) [reply test/e2e/framework/framework.go]: good to know!

**Comment by fanminshi** [comment]
close this in favor of https://github.com/coreos/etcd-operator/pull/1591

**Pull request closed without merging**